File: cpp/92.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
         int index = 1; 
         ListNode *newHead = new ListNode(-1);
         newHead->next = head;
         ListNode *t = newHead;
         while(index < m){
             t = t->next;
             index++;
         }
         ListNode *pre = t->next;
         ListNode *cur = t->next->next;
         index++;
         while(index <= n && cur){
             ListNode *tmpNode = cur->next;
             cur->next = pre;
             pre = cur;
             cur = tmpNode;
             index++;
         }
         t->next->next = cur;
         t->next = pre;
         return newHead->next;
    }
};

File: golang/92/92.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if left == 1 {
		return reverseN(head, right)
	}

	head.Next = reverseBetween(head.Next, left-1, right-1)
	return head
}

var successor *ListNode

func reverseN(head *ListNode, n int) *ListNode {
	if n == 1 {
		successor = head.Next
		return head
	}

	last := reverseN(head.Next, n-1)
	head.Next.Next = head
	head.Next = successor
	return last
}
